### DISCUSSION TITLE
How to ADB into Pi?

### DISCUSSION THREAD

**Discussion by drderiv** [tesla-android/issue-tracker#275] [Q&A] [answered]
I (and I'm sure others) would appreciate it if someone could post how to adb into a Pi running Tesla Android from another device (e.g., a Windows computer).  Thanks in advance.

**Comment by mikegapinski** [comment]
If you use a Pi 4 you can just connect the usb c cable to it, no permission dialogs etc:

<img width="336" alt="image" src="https://github.com/tesla-android/issue-tracker/assets/930055/340ecbd9-5add-4484-9664-ff8bb6e918d7">

You might need some generic drivers from google that come with the Android SDK on Windows. 

Once you are done with the connection you can enable wireless adb:
adb root
setprop service.adb.tcp.port 5555
stop adbd
start adbd

adb connect ***wifi.ip.address***:5555

Note: You have to be connected to the Tesla Android Wi-Fi network in order to connect wirelessly. 

I am going to harden things a little in the next release to make it more "production like". adb root will no longer be available. Connecting via USB will require accepting the permission dialog and wireless access would be configurable in the Tesla Android frontend. 

Another note: CM4 has only wireless or serial for ADB, that is why I am going to make wireless easier to configure and setup

**Reply by drderiv** [reply]
Thanks for the very complete answer.

I couldn't get past the first step, which is why I posted the question, and wondered if I was doing something wrong.  My Windows computer has USB C, but when I attach to the Pi 4's USB C, Windows isn't recognizing that anything is connected to it (though it is powering the Pi).  The Pi boots, I can attach to its hotspot network from the same Windows computer, and all is well, other than the fact that I can't get Windows to even see that something is connected via USB C.  I have the generic drivers available, but first Windows need to recognize that it's got a connection that needs them.

For what it's worth, on the Pi, the Developer option for USB debugging is on, the one for Wireless debugging is off (and trying to turn it on gives a message that says "Please connect to a Wi-Fi network", then the setting turns itself off).  Also the Default USB configuration setting was set to "No data transfer" originally (I think), and I've tried setting it to "USB tethering", but that doesn't make any difference.   I would have thought that just making the hardware connection by plugging in the USB C cable would have triggered a response from Windows Device Manager that something was connected, so none of these settings would matter, at least initially.

Any thoughts as to what the issue might be?

**Reply by mikegapinski** [reply]
Wireless debugging in the settings can be enabled only when you are connected to the network. Tesla Android acts as a router in this case, that is why it is not available. The commands I pasted perform the same action as the switch in the developer settings.

I remember that in the early days of Tesla Android people on windows had issues with fastboot. The drivers are in the same bundle so this problem should be solvable with the same solution. There you go: https://github.com/tesla-android/tesla-android.github.io/blob/main/install-guide-2022-18-1.md

You need the most current platform tools and the most current drivers for the Android 13 API level. I use the newest OS.

If you cannot see the device in device manager this could mean the cable you are using is either charge only or does not have all the pins that are needed. USB-C on windows machines is still meh. Try a USB-A to C cable. Those always work, even when they are USB 2.0 only internally.

BTW: If you are hoping to do adb root and adb remount to change something on the system/vendor partitions OTA will not work. Overlayfs is a no-go for seamless OTA

**Reply by drderiv** [reply]
Thanks again.  The same cable works when plugging into my Android phone.  The Windows computer recognizes it and "adb devices" returns as expected.  I'll check out the post you linked to and see if I can make any progress.

**Reply by mikegapinski** [reply, marked as answer]
> Thanks again. The same cable works when plugging into my Android phone. The Windows computer recognizes it and "adb devices" returns as expected. I'll check out the post you linked to and see if I can make any progress.

The Pi needs power delivery, it might not work if the usb port cannot deliver 2A(3 is recommended) on that port. Random stuff is happening with the board when it does not get enough juce. The phone is fine with 500mA.. 

I am sure it works though, I am doing tons of adb on those builds every day

**Reply by drderiv** [reply]
You were spot on.  I was able to connect via adb (not wirelessly, but that's fine; even when I stopped and started adbd with a single command joined with a "&&" I got dropped and had to reboot).  The primary issue was essentially the "cable" not having all the pins.  My actual cable was fine, but I had my pi mounted in an Argon ONE V2 case and the case stops access to the pi's built-in USB C and provides its own on the back of the unit.  That connector apparently functions as if it's power only.  Hopefully this helps someone else trying to use that case in the future.

Thanks for helping troubleshoot.  The reason I wanted adb access was to try to get the i2c fan in the case going, as you discussed [here](https://github.com/tesla-android/issue-tracker/issues/243).  i2cget and i2cset appear to be in the build, and they run, but there is no /dev/i2c* so it throws an error.  Anyhow, given the cruddy USB C support, I'm going to switch cases and go with an always-on fan, which solves all of my current issues.

Great project.  Thanks for doing it!

**Reply by mikegapinski** [reply]
> You were spot on. I was able to connect via adb (not wirelessly, but that's fine; even when I stopped and started adbd with a single command joined with a "&&" I got dropped and had to reboot). The primary issue was essentially the "cable" not having all the pins. My actual cable was fine, but I had my pi mounted in an Argon ONE V2 case and the case stops access to the pi's built-in USB C and provides its own on the back of the unit. That connector apparently functions as if it's power only. Hopefully this helps someone else trying to use that case in the future.
> 
> Thanks for helping troubleshoot. The reason I wanted adb access was to try to get the i2c fan in the case going, as you discussed [here](https://github.com/tesla-android/issue-tracker/issues/243). i2cget and i2cset appear to be in the build, and they run, but there is no /dev/i2c* so it throws an error. Anyhow, given the cruddy USB C support, I'm going to switch cases and go with an always-on fan, which solves all of my current issues.
> 
> Great project. Thanks for doing it!

The Argon One case is made of metal and blocks wifi signal/degrades performance. Not a total blocker if your car is a 2021+.

Anything that is either plastic or has tons of ventilation holes would do. Get the cheapest cooler from waveshare, they even come with PWM. Avoid those small "copper" radiators, those are hot garbage.
